### PR TITLE
added support for asg target-lifecycle-state path

### PR DESCRIPTION
**Pull request by Lu-David**
Issue #218

Description of changes:
1) Added core functionality for handling /latest/meta-data/autoscaling/target-lifecycle-state GET requests and changing state after set delay or trigger time. 
2) Updated README
3) Added unit tests for asglifecycle subcommand
4) Added e2e tests to test that asglifecycle target-lifecycle-state changes from InService to Terminated after delay

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/mock/asglifecycle/asglifecycle.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	"net/http"
+	"strings"
+	"time"
+
+	cfg "github.com/aws/amazon-ec2-metadata-mock/pkg/config"
+	"github.com/aws/amazon-ec2-metadata-mock/pkg/server"
+)
+
+const (
+	targetLifeCycleStatePath = "/latest/meta-data/autoscaling/target-lifecycle-state"
```
Comment (Lu-David) [inline pkg/mock/asglifecycle/asglifecycle.go:27]: Replace with the path in Config. 

E.g. c.Metadata.Paths.ASGLifecycle

File: pkg/mock/asglifecycle/asglifecycle.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+var (
+	eligibleIPs          = make(map[string]bool)
+	c                    cfg.Config
+	autoscalingStartTime int64 = time.Now().Unix()
+	state                      = "InService"
```
Comment (Lu-David) [inline pkg/mock/asglifecycle/asglifecycle.go:34]: Put Terminated and InService in Config

File: pkg/cmd/root/root.go [hunk 3]
```diff
@@ -43,14 +44,16 @@ var (
...
-		gf.MockTriggerTimeFlag:      "",
-		gf.MockIPCountFlag:          2,
-		gf.SaveConfigToFileFlag:     false,
-		gf.Imdsv2Flag:               false,
-		gf.RebalanceDelayInSecFlag:  0,
-		gf.RebalanceTriggerTimeFlag: "",
+		gf.ConfigFileFlag:                cfg.GetDefaultCfgFileName(),
+		gf.MockDelayInSecFlag:            0,
+		gf.MockTriggerTimeFlag:           "",
+		gf.MockIPCountFlag:               3,
```
Comment (Lu-David) [inline pkg/cmd/root/root.go:50]: Change the default value back to 2. No need to change it

File: test/e2e/cmd/asglifecycle-test [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  delay_in_sec=$2
+  tput setaf $MAGENTA
+  health_check $LIFECYCLE_STATE_TEST_PATH
+  TOKEN=$(get_v2Token $MAX_TOKEN_TTL $AEMM_PORT)
+
+  expected_state="InService"
+  response=$(curl -s -H "X-aws-ec2-metadata-token: $TOKEN" $LIFECYCLE_STATE_TEST_PATH)
+  assert_value "$response" "$expected_state" "test_asglifecycle::asg-pre-delay_response"
+
+  # ensure no impact to mock-delay functionality
```
Comment (Lu-David) [inline test/e2e/cmd/asglifecycle-test:33]: nit: change mock-delay to asg-termination-delay

File: pkg/mock/asglifecycle/asglifecycle.go (hunk 1 above)
Comment (Lu-David) [reply pkg/mock/asglifecycle/asglifecycle.go:27]: Actually, to keep it consistent with the other mocks, I'll keep this as is

File: pkg/mock/asglifecycle/asglifecycle.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+// Handler processes http requests
+func Handler(res http.ResponseWriter, req *http.Request) {
+	if c.MockIPCount >= 0 {
+		// req.RemoteAddr is formatted as IP:port
+		requestIP := strings.Split(req.RemoteAddr, ":")[0]
+		if !eligibleIPs[requestIP] {
+			if len(eligibleIPs) < c.MockIPCount {
+				eligibleIPs[requestIP] = true
+			} else {
+				log.Printf("Requesting IP %s is not eligible for Spot ITN or Rebalance Recommendation because the max number of IPs configured (%d) has been reached.\n", requestIP, c.MockIPCount)
```
Comment (Lu-David) [inline pkg/mock/asglifecycle/asglifecycle.go:56]: Change this log output

File: pkg/cmd/root/root.go (hunk 3 above)
Comment (LikithaVemulapalli) [reply pkg/cmd/root/root.go:50]: For all the comments that you added for this PR, are you planning to make those changes or already have those changes? I'm little bit confused with your comments

File: README.md [hunk 6]
```diff
@@ -138,23 +139,26 @@ Examples:
...
-  -d, --mock-delay-sec int              spot itn delay in seconds, relative to the application start time (default: 0 seconds)
-  -x, --mock-ip-count int               number of IPs in a cluster that can receive a Spot Interrupt Notice and/or Scheduled Event (default 2)
-      --mock-trigger-time string        spot itn trigger time in RFC3339 format. This takes priority over mock-delay-sec (default: none)
-  -p, --port string                     the HTTP port where the mock runs (default: 1338)
-      --rebalance-delay-sec int         rebalance rec delay in seconds, relative to the application start time (default: 0 seconds)
-      --rebalance-trigger-time string   rebalance rec trigger time in RFC3339 format. This takes priority over rebalance-delay-sec (default: none)
-  -s, --save-config-to-file             whether to save processed config from all input sources in .ec2-metadata-mock/.aemm-config-used.json in $HOME or working dir, if homedir is not found (default: false)
-      --version                         version for ec2-metadata-mock
+  -g, --asg-termination-delay-sec int      asg termination delay in seconds, relative to the application start time (default: 0 seconds)
```
Comment (LikithaVemulapalli) [inline README.md:148]: You randomly picked `g` or is it defined in the docs?

File: pkg/cmd/root/root.go [hunk 7]
```diff
@@ -89,9 +92,11 @@ func NewCmd() *cobra.Command {
 	cmd.PersistentFlags().BoolP(gf.Imdsv2Flag, "I", false, "whether to enable IMDSv2 only, requiring a session token when submitting requests (default: false, meaning both IMDS v1 and v2 are enabled)")
 	cmd.PersistentFlags().Int64(gf.RebalanceDelayInSecFlag, 0, "rebalance rec delay in seconds, relative to the application start time (default: 0 seconds)")
 	cmd.PersistentFlags().String(gf.RebalanceTriggerTimeFlag, "", "rebalance rec trigger time in RFC3339 format. This takes priority over "+gf.RebalanceDelayInSecFlag+" (default: none)")
+	cmd.PersistentFlags().Int64P(gf.ASGTerminationDelayInSecFlag, "g", 0, "asg termination delay in seconds, relative to the application start time (default: 0 seconds)")
+	cmd.PersistentFlags().Int64P(gf.ASGTerminationTriggerTimeFlag, "", 0, "asg termination trigger time in RFC3339 format. This takes priority over "+gf.ASGTerminationDelayInSecFlag+" (default: none)")
 
 	// add subcommands
-	cmd.AddCommand(spot.Command, events.Command)
+	cmd.AddCommand(spot.Command, events.Command, asglifecycle.Command)
```
Comment (LikithaVemulapalli) [inline pkg/cmd/root/root.go:99]: can we have `asgLifecycleTermination` throughout to maintain consistency excluding the places were we define the kind to be asglifecycle?

**Review by LikithaVemulapalli — CHANGES_REQUESTED** [review]
added few comments

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Lu-David) [reply README.md:148]: it was the only short hand flag available. All other flags were being used

File: pkg/cmd/cmdutil/cmdutil.go [hunk 8]
```diff
@@ -110,18 +111,22 @@ func getHandlerPairs(cmd *cobra.Command, config cfg.Config) []handlerPair {
 
 	isSpot := strings.Contains(cmd.Name(), "spot")
 	isEvents := strings.Contains(cmd.Name(), "events")
+	isASGLifecycle := strings.Contains(cmd.Name(), "asglifecycle")
```
[the remark this answers is not in the document]
Comment (Lu-David) [reply pkg/cmd/cmdutil/cmdutil.go:114]: Sounds good. Will rename all variables to be ASGLifecycleTermination

File: README.md [hunk 9]
```diff
@@ -394,6 +403,45 @@ $ curl localhost:1338/latest/meta-data/events/maintenance/scheduled
...
+$ ec2-metadata-mock asglifecycle --help
+Mock EC2 Auto Scaling Group Lifecycle Termination State
+
+Usage:
+  ec2-metadata-mock asglifecycle
+
+Aliases:
+  asglifecycle, autoscaling, asg
+
+Examples:
```
[the remark this answers is not in the document]
Comment (Lu-David) [reply README.md:419]: Sounds good. I'll delete this entire section

[the remark this answers is not in the document]
Comment (Lu-David) [reply README.md:419]: Lines 409-437

File: README.md [hunk 10]
```diff
@@ -394,6 +403,45 @@ $ curl localhost:1338/latest/meta-data/events/maintenance/scheduled
 }
 ```
 
+## Auto Scaling Group Lifecycle Termination
+Similar to spot, the `asglifecycle` command, view the local flags using `asglifecycle --help`:
```
[the remark this answers is not in the document]
Comment (Lu-David) [reply README.md:407]: ack

File: pkg/cmd/root/root.go (hunk 7 above)
Comment (Lu-David) [reply pkg/cmd/root/root.go:99]: in this line you highlighted, the asglifecycle is of kind asglifecycle. I can't rename it without renaming all packages to asglifecycletermination. I'll keep this one as is if that's okay?

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (LikithaVemulapalli) [reply README.md:419]: no, I want you to remove only examples: were it is repetitive not entire section, you can have lines 410-418 and 439-444

File: pkg/cmd/root/root.go (hunk 7 above)
Comment (LikithaVemulapalli) [reply pkg/cmd/root/root.go:99]: yeah you can keep it, and modify other places wherever it is possible, thanks

Comment (Lu-David) [reply pkg/cmd/root/root.go:99]: Resolution: keeping everything as asgLifecycle until Launch is introduced. When launch is introduced, will rename

File: pkg/cmd/cmdutil/cmdutil.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Lu-David) [reply pkg/cmd/cmdutil/cmdutil.go:114]: Resolution: keeping everything as just asglifecycle for consistency

**Review by LikithaVemulapalli — APPROVED** [review]
/lgtm

**Pull request merged**